webapp/src/blocks.tsx

```diff
@@ -891,14 +892,77 @@ export class Editor extends toolboxeditor.ToolboxEditor {
         this.parent.fireResize();
     }
 
-    listenToBlockErrorChanges(handlerKey: string, handler: (errors: pxtblockly.BlockDiagnostic[]) => void) {
-        this.errorChangesListeners[handlerKey] = handler;
+    onExceptionDetected(exception: pxsim.DebuggerBreakpointMessage) {
+        const displayInfo: ErrorDisplayInfo = this.getDisplayInfoForException(exception);
+        this.errors = [displayInfo];
+    }
+
+    private onBlockErrorChanges(errors: ErrorDisplayInfo[]) {
+        this.errors = errors;
+    }
+
+    private getDisplayInfoForBlockError(blockId: string, message: string): ErrorDisplayInfo {
+        return {
+            message: message,
+            onClick: () => {
+                // Block may already be highlighted, so we want to set highlighted directly to true
+                // as opposed to using `highlightBlock`, which toggles.
+                const block = this.editor.getBlockById(blockId);
+                if (!block) {
+                    return;
+                }
+
+                block.setHighlighted(true);
+                this.editor.centerOnBlock(blockId);
+            }
+        }
     }
 
-    private onBlockErrorChanges(errors: pxtblockly.BlockDiagnostic[]) {
-        for (let listener of pxt.U.values(this.errorChangesListeners)) {
-            listener(errors)
+    private getDisplayInfoForException(exception: pxsim.DebuggerBreakpointMessage): ErrorDisplayInfo {
+        const message = pxt.Util.rlf(exception.exceptionMessage);
+        const stackFrames: StackFrameDisplayInfo[] = exception.stackframes?.map(frame => {
+            const locInfo = frame.funcInfo as pxtc.FunctionLocationInfo;
+            if (!locInfo?.functionName) {
+                return undefined;
+            }
+
+            const blockId = this.compilationResult ? pxtblockly.findBlockIdByLine(this.compilationResult.sourceMap, { start: locInfo.line, length: locInfo.endLine - locInfo.line }) : undefined;
+            if (!blockId) {
+                return undefined;
+            }
+
+            // Get human-readable block text
+            const block = this.editor.getBlockById(blockId);
+            if (!block) {
+                return undefined;
+            }
+            const blockText = this.getBlockText(block);
+
+            return {
+                message: blockText ? lf("at the '{0}' block", blockText) : lf("at {0}", locInfo.functionName),
+                onClick: () => this.highlightStatement(locInfo, exception)
+            };
+        }).filter(f => !!f) ?? undefined;
+
+        return {
+            message,
+            stackFrames
+        };
+    }
+
+    private getBlockText(block: Blockly.BlockSvg): string {
+        const fieldValues = [];
+        for (const input of block.inputList) {
+            if (input.fieldRow.length > 0) {
+                for (const field of input.fieldRow) {
+                    const text = field.getText();
```

Comment:
this isn't going to work well for all fields; i'm guessing that assets are just going to produce some un-pronounceable nonsense